In [2]:
import numpy as np
import pandas as pd
heroes = pd.read_csv("data/heroes_information.csv")

In [3]:
###instalar desde anaconda navigator######
from pandasql import *

In [4]:
pysqldf = lambda q: sqldf(q,globals())

In [19]:
query2 = " SELECT * FROM heroes"

In [20]:
pysqldf(query2)

id             name  Gender Eye color               Race  \
0      0           A-Bomb    Male    yellow              Human   
1      1       Abe Sapien    Male      blue      Icthyo Sapien   
2      2         Abin Sur    Male      blue            Ungaran   
3      3      Abomination    Male     green  Human / Radiation   
4      4          Abraxas    Male      blue      Cosmic Entity   
..   ...              ...     ...       ...                ...   
729  729  Yellowjacket II  Female      blue              Human   
730  730             Ymir    Male     white        Frost Giant   
731  731             Yoda    Male     brown     Yoda's species   
732  732          Zatanna  Female      blue              Human   
733  733             Zoom    Male       red               None   

           Hair color  Height          Publisher Skin color Alignment  Weight  
0             No Hair   203.0      Marvel Comics       None      good   441.0  
1             No Hair   191.0  Dark Horse Comics       blue      good    65.0  
2             No Hair   185.0          DC Comics        red      good    90.0  
3             No Hair   203.0      Marvel Comics       None       bad   441.0  
4               Black   -99.0      Marvel Comics       None       bad   -99.0  
..                ...     ...                ...        ...       ...     ...  
729  Strawberry Blond   165.0      Marvel Comics       None      good    52.0  
730           No Hair   304.8      Marvel Comics      white      good   -99.0  
731             White    66.0       George Lucas      green      good    17.0  
732             Black   170.0          DC Comics       None      good    57.0  
733             Brown   185.0          DC Comics       None       bad    81.0  

[734 rows x 11 columns]

# 1) ¿Cuál es el nombre del primer superhéroe de la tabla heroes?

In [31]:
query_1 = " SELECT * FROM heroes"
pysqldf(query_1)

id             name  Gender Eye color               Race  \
0      0           A-Bomb    Male    yellow              Human   
1      1       Abe Sapien    Male      blue      Icthyo Sapien   
2      2         Abin Sur    Male      blue            Ungaran   
3      3      Abomination    Male     green  Human / Radiation   
4      4          Abraxas    Male      blue      Cosmic Entity   
..   ...              ...     ...       ...                ...   
729  729  Yellowjacket II  Female      blue              Human   
730  730             Ymir    Male     white        Frost Giant   
731  731             Yoda    Male     brown     Yoda's species   
732  732          Zatanna  Female      blue              Human   
733  733             Zoom    Male       red               None   

           Hair color  Height          Publisher Skin color Alignment  Weight  
0             No Hair   203.0      Marvel Comics       None      good   441.0  
1             No Hair   191.0  Dark Horse Comics       blue      good    65.0  
2             No Hair   185.0          DC Comics        red      good    90.0  
3             No Hair   203.0      Marvel Comics       None       bad   441.0  
4               Black   -99.0      Marvel Comics       None       bad   -99.0  
..                ...     ...                ...        ...       ...     ...  
729  Strawberry Blond   165.0      Marvel Comics       None      good    52.0  
730           No Hair   304.8      Marvel Comics      white      good   -99.0  
731             White    66.0       George Lucas      green      good    17.0  
732             Black   170.0          DC Comics       None      good    57.0  
733             Brown   185.0          DC Comics       None       bad    81.0  

[734 rows x 11 columns]

**Respuesta:** A-Bomb

# 2) ¿Cuántas casas publicadoras existen?

In [44]:
query_2 = "SELECT COUNT(DISTINCT Publisher) FROM heroes"
pysqldf(query_2)

COUNT(DISTINCT Publisher)
0                         24

**Respuesta:** 24

# 3) ¿Cuántos superhéroes miden más de 2 metros (200 cms)

In [147]:
query_3 = "SELECT COUNT(*) FROM heroes WHERE Height > 200"
pysqldf(query_3)

COUNT(*)
0        59

**Respuesta:** 59

# 4) ¿Cuántos superhéroes son humanos y miden más de 2 metros (200 cms)?

In [150]:
query_4 = "SELECT COUNT(*) FROM heroes WHERE Height > 200 AND Race = 'Human'"
pysqldf(query_4)

COUNT(*)
0        12

**Respuesta:** 12

# 5 ¿Cuántos superhéroes pesan más de 100 lbs y menos de 200 lbs?

In [151]:
query_5 = "SELECT Count(*) FROM heroes WHERE Weight > 100 AND Weight < 200 "
pysqldf(query_5)

Count(*)
0        98

**Respuesta:** 98

# 6) ¿Cuántos superhéroes tienen los ojos de color azul o rojo?

In [146]:
query_6 = "SELECT Count(*) FROM heroes WHERE `Eye color` = 'red' OR `Eye color` = 'blue'"
pysqldf(query_6)

Count(*)
0       271

**Respuesta:** 271

# 7) ¿Cuántos superhéroes son Human, Mutant y tienen el pelo color Verde o son Vampiros con pelo Negro?

In [165]:
query_7 = " SELECT COUNT(*) FROM heroes WHERE  (Race IN ('Human','Mutant') AND `Hair color` = 'Green') OR  (Race = 'Vampire' AND `Hair color` = 'Black')"
pysqldf(query_7)

COUNT(*)
0         5

**Respuesta:** 5

# 8) ¿Cuál es el nombre del primer superhéroe si ordeno la tabla por raza en orden descendente?

In [167]:
query_8 = " SELECT * FROM heroes ORDER BY Race DESC"
pysqldf(query_8)

id            name  Gender Eye color             Race Hair color  \
0    611  Solomon Grundy    Male     black           Zombie      White   
1    274          Gamora  Female    yellow    Zen-Whoberian      Black   
2    731            Yoda    Male     brown   Yoda's species      White   
3    527        Predator    Male      None           Yautja       None   
4     17           Alien    Male      None  Xenomorph XX121    No Hair   
..   ...             ...     ...       ...              ...        ...   
729  723          Wondra  Female      None             None       None   
730  724  Wyatt Wingfoot    Male     brown             None      Black   
731  726           X-Man    Male      blue             None      Brown   
732  727     Yellow Claw    Male      blue             None    No Hair   
733  733            Zoom    Male       red             None      Brown   

     Height          Publisher Skin color Alignment  Weight  
0     279.0          DC Comics       None       bad   437.0  
1     183.0      Marvel Comics      green      good    77.0  
2      66.0       George Lucas      green      good    17.0  
3     213.0  Dark Horse Comics       None       bad   234.0  
4     244.0  Dark Horse Comics      black       bad   169.0  
..      ...                ...        ...       ...     ...  
729   -99.0      Marvel Comics       None      good   -99.0  
730   196.0      Marvel Comics       None      good   117.0  
731   175.0      Marvel Comics       None      good    61.0  
732   188.0      Marvel Comics       None       bad    95.0  
733   185.0          DC Comics       None       bad    81.0  

[734 rows x 11 columns]

**Respuesta:** Solomon Grundy

# 9) ¿Cuántos superhéroes son de género masculino y cuántas de género femenino?

In [168]:
query_9 = "SELECT Gender,COUNT(*) FROM heroes GROUP BY Gender" # `Hair color`
pysqldf(query_9)

Gender  COUNT(*)
0    None        29
1  Female       200
2    Male       505

**Respuesta:** Female = 200 &nbsp; | &nbsp; Male = 505

# 10) ¿Cuántas casas publicadoras tienen más de 15 superhéroes?

In [191]:
query_10 = " SELECT COUNT(*) FROM (SELECT COUNT(*) FROM heroes GROUP BY Publisher HAVING COUNT(*) > 15)"
pysqldf(query_10)

COUNT(*)
0         4

**Respuesta:** 4